File: backend/src/main/java/br/ufpr/sept/so2/modules/bff/infrastructure/SolicitacoesDashboardQueryAdapter.java
package br.ufpr.sept.so2.modules.bff.infrastructure;

import br.ufpr.sept.so2.modules.bff.application.ports.SolicitacoesDashboardQueryPort;
import br.ufpr.sept.so2.modules.solicitacoes.application.ports.SolicitacaoRepository;
import br.ufpr.sept.so2.modules.solicitacoes.domain.Solicitacao;
import org.springframework.data.domain.PageRequest;
import org.springframework.data.domain.Sort;
import org.springframework.stereotype.Component;
import org.springframework.transaction.annotation.Transactional;

import java.time.OffsetDateTime;
import java.util.List;
import java.util.Set;
import java.util.UUID;

@Component
public class SolicitacoesDashboardQueryAdapter implements SolicitacoesDashboardQueryPort {

    private static final Set<String> ESTADOS_ABERTOS = Set.of("EM_ANALISE", "EM_AJUSTE");
    private static final String ESTADO_PENDENCIA = "EM_AJUSTE";

    private final SolicitacaoRepository solicitacaoRepository;

    public SolicitacoesDashboardQueryAdapter(SolicitacaoRepository solicitacaoRepository) {
        this.solicitacaoRepository = solicitacaoRepository;
    }

    @Override
    @Transactional(readOnly = true)
    public SolicitacoesDashboard consultar(UUID solicitanteId) {
        Sort recente = Sort.by(Sort.Direction.DESC, "createdAt");
        OffsetDateTime agora = OffsetDateTime.now();
        List<SolicitacaoResumo> ultimas = solicitacaoRepository
                .findMinhas(solicitanteId, null, null, null, PageRequest.of(0, 5, recente))
                .getContent()
                .stream()
                .map(item -> toResumo(item, agora))
                .toList();
        List<PendenciaResumo> pendencias = solicitacaoRepository
                .findMinhas(solicitanteId, ESTADO_PENDENCIA, null, null, PageRequest.of(0, 3, recente))
                .getContent()
                .stream()
                .map(SolicitacoesDashboardQueryAdapter::toPendencia)
                .toList();
        return new SolicitacoesDashboard(contarAbertas(solicitanteId), pendencias, ultimas);
    }

    private int contarAbertas(UUID solicitanteId) {
        int total = 0;
        for (String estado : ESTADOS_ABERTOS) {
            total += (int) solicitacaoRepository
                    .findMinhas(solicitanteId, estado, null, null, PageRequest.of(0, 1))
                    .getTotalElements();
        }
        return total;
    }

    private static SolicitacaoResumo toResumo(Solicitacao solicitacao, OffsetDateTime agora) {
        return new SolicitacaoResumo(
                solicitacao.getId(),
                solicitacao.getProtocolo().getValor(),
                solicitacao.getTipoNome(),
                solicitacao.getEstado(),
                solicitacao.getPrazoEm(),
                solicitacao.prazoVencido(agora)
        );
    }

    private static PendenciaResumo toPendencia(Solicitacao solicitacao) {
        return new PendenciaResumo(
                solicitacao.getId(),
                solicitacao.getTipoNome(),
                solicitacao.getEstado(),
                "/solicitacoes/" + solicitacao.getId()
        );
    }
}

File: backend/src/main/kotlin/br/ufpr/sept/so2/modules/bff/infrastructure/SolicitacoesDashboardQueryAdapter.kt
package br.ufpr.sept.so2.modules.bff.infrastructure

import br.ufpr.sept.so2.modules.bff.application.ports.SolicitacoesDashboardQueryPort
import br.ufpr.sept.so2.modules.bff.application.ports.SolicitacoesDashboardQueryPort.PendenciaResumo
import br.ufpr.sept.so2.modules.bff.application.ports.SolicitacoesDashboardQueryPort.SolicitacaoResumo
import br.ufpr.sept.so2.modules.bff.application.ports.SolicitacoesDashboardQueryPort.SolicitacoesDashboard
import br.ufpr.sept.so2.modules.solicitacoes.application.ports.SolicitacaoRepository
import br.ufpr.sept.so2.modules.solicitacoes.domain.Solicitacao
import org.springframework.data.domain.PageRequest
import org.springframework.data.domain.Sort
import org.springframework.stereotype.Component
import org.springframework.transaction.annotation.Transactional
import java.time.OffsetDateTime
import java.util.UUID

@Component
class SolicitacoesDashboardQueryAdapter(
    private val solicitacaoRepository: SolicitacaoRepository,
) : SolicitacoesDashboardQueryPort {

    @Transactional(readOnly = true)
    override fun consultar(solicitanteId: UUID): SolicitacoesDashboard {
        val recente = Sort.by(Sort.Direction.DESC, "createdAt")
        val agora = OffsetDateTime.now()
        val ultimas = solicitacaoRepository
            .findMinhas(solicitanteId, null, null, null, PageRequest.of(0, 5, recente))
            .content
            .map { toResumo(it, agora) }
        val pendencias = solicitacaoRepository
            .findMinhas(solicitanteId, ESTADO_PENDENCIA, null, null, PageRequest.of(0, 3, recente))
            .content
            .map(::toPendencia)
        return SolicitacoesDashboard(contarAbertas(solicitanteId), pendencias, ultimas)
    }

    private fun contarAbertas(solicitanteId: UUID): Int =
        ESTADOS_ABERTOS.sumOf { estado ->
            solicitacaoRepository
                .findMinhas(solicitanteId, estado, null, null, PageRequest.of(0, 1))
                .totalElements
                .toInt()
        }

    companion object {
        private val ESTADOS_ABERTOS = setOf("EM_ANALISE", "EM_AJUSTE")
        private const val ESTADO_PENDENCIA = "EM_AJUSTE"

        private fun toResumo(solicitacao: Solicitacao, agora: OffsetDateTime): SolicitacaoResumo =
            SolicitacaoResumo(
                solicitacao.id,
                solicitacao.protocolo.valor,
                solicitacao.tipoNome,
                solicitacao.estado,
                solicitacao.prazoEm,
                solicitacao.prazoVencido(agora),
            )

        private fun toPendencia(solicitacao: Solicitacao): PendenciaResumo =
            PendenciaResumo(
                solicitacao.id,
                solicitacao.tipoNome,
                solicitacao.estado,
                "/solicitacoes/${solicitacao.id}",
            )
    }
}
